app/src/androidTest/java/com/sdp13epfl2021/projmag/database/CachedProjectsDatabaseTest.kt

```diff
@@ -226,91 +226,4 @@ class CachedProjectsDatabaseTest {
         assertEquals(4, count.get())
     }
 
-
-
-
-    private class FakeDatabase(projectsBeginning: List<ImmutableProject>) : ProjectsDatabase {
-        private var projects: List<ImmutableProject> = projectsBeginning
-        private var listeners: List<((ProjectChange) -> Unit)> = emptyList()
-        private var nextId: Int = 90000
-
-        fun add(project: ImmutableProject) {
-            projects = projects + project
-            listeners.forEach{ it -> it(ProjectChange(ProjectChange.Type.ADDED, project)) }
-        }
-
-        fun modify(project: ImmutableProject) {
-            projects.filter { p -> p.id != project.id }
-            projects = projects + project
-            listeners.forEach{ it -> it(ProjectChange(ProjectChange.Type.MODIFIED, project)) }
-        }
-
-        fun remove(project: ImmutableProject) {
-            projects = projects - project
-            listeners.forEach{ it -> it(ProjectChange(ProjectChange.Type.REMOVED, project)) }
-        }
-
-        override fun getAllIds(
-            onSuccess: (List<ProjectId>) -> Unit,
-            onFailure: (Exception) -> Unit
-        ) { }
-
-        override fun getProjectFromId(
-            id: ProjectId,
-            onSuccess: (ImmutableProject?) -> Unit,
-            onFailure: (Exception) -> Unit
-        ) { }
-
-        override fun getAllProjects(
-            onSuccess: (List<ImmutableProject>) -> Unit,
-            onFailure: (Exception) -> Unit
-        ) {
-            onSuccess(projects)
-        }
-
-        override fun getProjectsFromName(
-            name: String,
-            onSuccess: (List<ImmutableProject>) -> Unit,
-            onFailure: (Exception) -> Unit
-        ) { }
-
-        override fun getProjectsFromTags(
-            tags: List<String>,
-            onSuccess: (List<ImmutableProject>) -> Unit,
-            onFailure: (Exception) -> Unit
-        ) { }
-
-        override fun pushProject(
-            project: ImmutableProject,
-            onSuccess: (ProjectId) -> Unit,
-            onFailure: (Exception) -> Unit
-        ) {
-            val pid = nextId.toString()
-            nextId += 1
-            val newProject = project?.let {ImmutableProject(pid,
-                it.name, it.lab, it.teacher, it.TA, it.nbParticipant, it.assigned, it.masterProject,
-                it.bachelorProject, it.tags, it.isTaken, it.description)
-            }
-            add(newProject)
-            onSuccess(pid)
-        }
-
-        override fun deleteProjectWithId(
-            id: ProjectId,
-            onSuccess: () -> Unit,
-            onFailure: (Exception) -> Unit
-        ) {
-            projects.find {p -> p.id == id}?.let { remove(it) }
-            onSuccess()
-        }
-
-        override fun addProjectsChangeListener(changeListener: (ProjectChange) -> Unit) {
-            listeners = listeners + changeListener
```

Comment:
Explain more explicitly by commenting what listeners are for and what they are associated to (because I thought there was one listener per project and it took me a while of reading your code to understand)

Reply:
Listeners on ProjectsDatabase are triggers at every projects change (i.e. on new project, on deleted project or on modified project).
The listeners are not associated to a particular project but on the whole collection of projects.